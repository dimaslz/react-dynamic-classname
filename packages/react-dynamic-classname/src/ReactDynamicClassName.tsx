import { FC } from 'react';
import { condition  } from './utils';

const resolveChildren = (children: any) => {
	if ('children' in children?.props) {
		const propChildren = children.props.children;
		const childrenType = propChildren.type;
		if (typeof childrenType === 'function') {
			children = propChildren.type(propChildren.props);
		}
	}

	if (typeof children.type === 'function') {
		children = children.type(children.props);
	}

	if (children) {
		children = (
			Array.isArray(children) ? children : [children]
		)
			.map((child: any, index: any) => {
				child = {
					...child,
					key: index,
				};

				return condition(child);
			},
			);
	}

	return children;
};

export const ReactDynamicClassName: FC = ({
	children,
}: any) => {
	if (Array.isArray(children)) {
		return children.map((c: any) => resolveChildren(c));
	}

	if (Array.isArray(children.props.children)) {
		return children.props.children.map((c: any) => resolveChildren(c));
	}

	return resolveChildren(children);
};

export default ReactDynamicClassName;
