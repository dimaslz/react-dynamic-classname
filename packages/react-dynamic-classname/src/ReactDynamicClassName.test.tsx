import { render } from '@testing-library/react';

import '@testing-library/jest-dom';

import ReactDynamicClassName from './ReactDynamicClassName';
import { FC } from 'react';

type PROPS = {
	children?: any | null,
	klass?: string | object | string[] | object[]
}

const DummyComponent: FC<PROPS> = ({ klass }: PROPS) => (
	<div className={JSON.stringify(klass)}>
		<div className={JSON.stringify(klass)}>
			<h3 className={JSON.stringify(klass)}>
				Hello Dummy!
			</h3>
		</div>
	</div>
);

const Demo: FC<PROPS> = ({ klass, children = null }: PROPS) =>
	 (<ReactDynamicClassName>
		<h1 className={JSON.stringify(klass)}>
			{children || 'Hello World!'}
		</h1>
	</ReactDynamicClassName>)
;

describe('ReactDynamicClassName', () => {
	describe('Using html elements as children', () => {
		test('use correct class from an Array', () => {
			const { baseElement, getByText } = render(<Demo klass={[{ 'class-a': true }, { 'class-b': false }]} />);
			const element = getByText('Hello World!');
			expect(element).toBeInTheDocument();

			expect(baseElement).toMatchSnapshot();
		});

		test('use correct class from an Object', () => {
			const { baseElement, getByText } = render(<Demo klass={{ 'class-a': true, 'class-b': false }} />);
			const element = getByText('Hello World!');
			expect(element).toBeInTheDocument();

			expect(baseElement).toMatchSnapshot();
		});

		test('use correct class from an Object and Array', () => {
			const { baseElement, getByText } = render(
				<Demo
					klass={[
						{ 'class-a': true, 'class-b': false },
						[{ 'class-c': true, 'class-d': false }],
					]}
				/>);
			const element = getByText('Hello World!');
			expect(element).toBeInTheDocument();

			expect(baseElement).toMatchSnapshot();
		});
	});

	describe('Using components as children', () => {
		test('components as children with class as array of objects', () => {
			const { baseElement, getByText } = render(<Demo>
				<DummyComponent klass={[{ 'class-a': true }, { 'class-b': false }]} />
			</Demo>);

			const element = getByText('Hello Dummy!');
			expect(element).toBeInTheDocument();

			expect(baseElement).toMatchSnapshot();
		});

		test('use correct class from an Object', () => {
			const { baseElement, getByText } = render(<Demo>
				<DummyComponent klass={{ 'class-a': true, 'class-b': false }} />
			</Demo>);

			const element = getByText('Hello Dummy!');
			expect(element).toBeInTheDocument();

			expect(baseElement).toMatchSnapshot();
		});

		test('use correct class from an Object and Array', () => {
			const { baseElement, getByText } = render(<Demo>
				<DummyComponent klass={[
					{ 'class-a': true, 'class-b': false },
					[{ 'class-c': true, 'class-d': false }],
				]}
				/>
			</Demo>);

			const element = getByText('Hello Dummy!');
			expect(element).toBeInTheDocument();

			expect(baseElement).toMatchSnapshot();
		});
	});

	describe('Using multiple components as children', () => {
		test('components as children with class as array of objects', () => {
			const { baseElement, getAllByText } = render(<Demo>
				<DummyComponent klass={[{ 'class-a': true }, { 'class-b': false }]} />
				<DummyComponent klass={[{ 'class-a': true }, { 'class-b': false }]} />
				<DummyComponent klass={[{ 'class-a': true }, { 'class-b': false }]} />
			</Demo>);

			const element = getAllByText('Hello Dummy!');
			expect(element).toHaveLength(3);

			expect(baseElement).toMatchSnapshot();
		});

		test('use correct class from an Object', () => {
			const { baseElement, getAllByText } = render(<Demo>
				<DummyComponent klass={{ 'class-a': true, 'class-b': false }} />
				<DummyComponent klass={{ 'class-a': true, 'class-b': false }} />
				<DummyComponent klass={{ 'class-a': true, 'class-b': false }} />
			</Demo>);

			const element = getAllByText('Hello Dummy!');
			expect(element).toHaveLength(3);

			expect(baseElement).toMatchSnapshot();
		});

		test('use correct class from an Object and Array', () => {
			const { baseElement, getAllByText } = render(<Demo>
				<DummyComponent klass={[
					{ 'class-a': true, 'class-b': false },
					[{ 'class-c': true, 'class-d': false }],
				]}
				/>
				<DummyComponent klass={[
					{ 'class-a': true, 'class-b': false },
					[{ 'class-c': true, 'class-d': false }],
				]}
				/>
				<DummyComponent klass={[
					{ 'class-a': true, 'class-b': false },
					[{ 'class-c': true, 'class-d': false }],
				]}
				/>
			</Demo>);

			const element = getAllByText('Hello Dummy!');
			expect(element).toHaveLength(3);

			expect(baseElement).toMatchSnapshot();
		});
	});

	describe('Using multiple elements as children', () => {
		test('elements as children with class as objects', () => {
			const { baseElement } = render(<Demo>
				<div className={JSON.stringify({ 'class-a': true, 'class-b': false })}>example A</div>
				<div className={JSON.stringify({ 'class-a': true, 'class-b': false })}>example B</div>
				<div className={JSON.stringify({ 'class-a': true, 'class-b': false })}>example C</div>
			</Demo>);

			expect(baseElement).toMatchSnapshot();
		});

		test('elements as children with class as array of objects', () => {
			const { baseElement } = render(<Demo>
				<div className={JSON.stringify([{ 'class-a': true }, { 'class-b': false }])}>example A</div>
				<div className={JSON.stringify([{ 'class-a': true }, { 'class-b': false }])}>example B</div>
				<div className={JSON.stringify([{ 'class-a': true }, { 'class-b': false }])}>example C</div>
			</Demo>);

			expect(baseElement).toMatchSnapshot();
		});

		test('elements as children with class as Array and Object', () => {
			const { baseElement } = render(<Demo>
				<div className={JSON.stringify([
					{ 'class-a': true, 'class-b': false },
					[{ 'class-c': true, 'class-d': false }],
				])}
				>example A</div>
				<div className={JSON.stringify([
					{ 'class-a': true, 'class-b': false },
					[{ 'class-c': true, 'class-d': false }],
				])}
				>example B</div>
				<div className={JSON.stringify([
					{ 'class-a': true, 'class-b': false },
					[{ 'class-c': true, 'class-d': false }],
				])}
				>example C</div>
			</Demo>);

			expect(baseElement).toMatchSnapshot();
		});
	});

	describe('Dummy', () => {
		test('ClassName native', () => {
			const { baseElement } = render(<ReactDynamicClassName>
				<div className="class-a">example A</div>
			</ReactDynamicClassName>);

			expect(baseElement).toMatchSnapshot();
		});
	});
});
