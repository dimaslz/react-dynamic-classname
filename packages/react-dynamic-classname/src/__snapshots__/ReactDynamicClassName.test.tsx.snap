// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactDynamicClassName Dummy ClassName native 1`] = `
<body>
  <div>
    <div
      class="class-a"
    >
      example A
    </div>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using components as children components as children with class as array of objects 1`] = `
<body>
  <div>
    <div
      class="class-a"
    >
      <div
        class="class-a"
      >
        <h3
          class="class-a"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using components as children use correct class from an Object 1`] = `
<body>
  <div>
    <div
      class="class-a"
    >
      <div
        class="class-a"
      >
        <h3
          class="class-a"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using components as children use correct class from an Object and Array 1`] = `
<body>
  <div>
    <div
      class="class-a class-c"
    >
      <div
        class="class-a class-c"
      >
        <h3
          class="class-a class-c"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using html elements as children use correct class from an Array 1`] = `
<body>
  <div>
    <h1
      class="class-a"
    >
      Hello World!
    </h1>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using html elements as children use correct class from an Object 1`] = `
<body>
  <div>
    <h1
      class="class-a"
    >
      Hello World!
    </h1>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using html elements as children use correct class from an Object and Array 1`] = `
<body>
  <div>
    <h1
      class="class-a class-c"
    >
      Hello World!
    </h1>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using multiple components as children components as children with class as array of objects 1`] = `
<body>
  <div>
    <div
      class="class-a"
    >
      <div
        class="class-a"
      >
        <h3
          class="class-a"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
    <div
      class="class-a"
    >
      <div
        class="class-a"
      >
        <h3
          class="class-a"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
    <div
      class="class-a"
    >
      <div
        class="class-a"
      >
        <h3
          class="class-a"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using multiple components as children use correct class from an Object 1`] = `
<body>
  <div>
    <div
      class="class-a"
    >
      <div
        class="class-a"
      >
        <h3
          class="class-a"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
    <div
      class="class-a"
    >
      <div
        class="class-a"
      >
        <h3
          class="class-a"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
    <div
      class="class-a"
    >
      <div
        class="class-a"
      >
        <h3
          class="class-a"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using multiple components as children use correct class from an Object and Array 1`] = `
<body>
  <div>
    <div
      class="class-a class-c"
    >
      <div
        class="class-a class-c"
      >
        <h3
          class="class-a class-c"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
    <div
      class="class-a class-c"
    >
      <div
        class="class-a class-c"
      >
        <h3
          class="class-a class-c"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
    <div
      class="class-a class-c"
    >
      <div
        class="class-a class-c"
      >
        <h3
          class="class-a class-c"
        >
          Hello Dummy!
        </h3>
      </div>
    </div>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using multiple elements as children elements as children with class as Array and Object 1`] = `
<body>
  <div>
    <div
      class="class-a class-c"
    >
      example A
    </div>
    <div
      class="class-a class-c"
    >
      example B
    </div>
    <div
      class="class-a class-c"
    >
      example C
    </div>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using multiple elements as children elements as children with class as array of objects 1`] = `
<body>
  <div>
    <div
      class="class-a"
    >
      example A
    </div>
    <div
      class="class-a"
    >
      example B
    </div>
    <div
      class="class-a"
    >
      example C
    </div>
  </div>
</body>
`;

exports[`ReactDynamicClassName Using multiple elements as children elements as children with class as objects 1`] = `
<body>
  <div>
    <div
      class="class-a"
    >
      example A
    </div>
    <div
      class="class-a"
    >
      example B
    </div>
    <div
      class="class-a"
    >
      example C
    </div>
  </div>
</body>
`;
